subsys/net/ip/tcp2.c

```diff
@@ -389,7 +394,9 @@ static int tcp_conn_unref(struct tcp *conn)
 	tcp_pkt_unref(conn->send_data);
 
 	if (CONFIG_NET_TCP_RECV_QUEUE_TIMEOUT) {
-		tcp_pkt_unref(conn->queue_recv_data);
+		if (conn->queue_recv_data) {
+			tcp_pkt_unref(conn->queue_recv_data);
+		}
```

Comment:
Isn't `conn->queue_recv_data` be always non-NULL? It is always allocated and if fails, then `tcp_conn_alloc()` cleanups and fails.

Reply:
@mniestroj: @jukkar made this change based on my earlier comment where I proposed to use this robust, context-free pattern (check if value initialized/deinitialize it). It's good that you performed "static analysis" and confirmed that **with the current code**, on all execution paths leading here, conn->queue_recv_data will be non-null. But can you guarantee that condition will be true for future evolution of the code? Obviously no. Can you guarantee that everyone who will patch this code will perform "static analysis"? I guess no. Can you guarantee that reviewers will account for the properties of the code (which even not visible in the patch?) I guess even less so.

So, "design patterns". (Robust context-free exception handling here). When this code is problem-free (wishful thinking) and not changed for 3-5 years, we can make a pass over it and remove any "superfluous" checks.

Reply:
> It is always allocated and if fails, then `tcp_conn_alloc()` cleanups and fails.

Indeed, good catch! The null check is not needed here.

Reply:
@pfalcon Usage of `conn->queue_recv_data` is really simple now. It is an allocated `net_pkt` if `tcp_conn_alloc()` was successful, then it is freed in `tcp_conn_unref()`. The former function is a constructor, while the latter is a destructor (if ref_count reaches zero). You basically don't call a destructor on an object that was not constructed properly.